src/server.ts

```diff
@@ -152,94 +118,50 @@ export class Server {
              *
              * Process Flow:
              * 1. Creates a new stream with a unique ID
-             * 2. Sets up storage for stream data and receivers
-             * 3. Forwards received chunks to all registered receivers
-             * 4. Cleans up resources when the stream ends
+             * 2. Notifies all instantiators that a stream message is pending
+             * 3. Awaits for each instantiator that has a reader for that channel to connect
+             * 4. Notify writing instantiator to start sending data
+             * 5. Forwards received chunks to all registered receivers
+             * 6. Cleans up resources when the stream ends
              */
             sendStreamMessage: async (
-                stream: grpc.ServerDuplexStream<DataChunk, Id>,
+                stream: grpc.ServerDuplexStream<StreamChunk, Id>,
             ) => {
-                const id = this.streamMsgId
-                this.streamMsgId += 1
-                this.logger.debug('Openin stream with id ' + id)
+                // Get the actual identifier of the channel
+                const identify = <StreamChunk>(
+                    await new Promise((res) => stream.once('data', res))
+                )
 
-                const obj: OpenStream = {
-                    done: [],
-                    receivers: [],
+                if (!identify.id) {
+                    throw 'no, expected first a stream identifier'
```

Comment:
Can we add a more descriptive error message that will help the pipeline developers?